File: 2024/day_01/python/main.py
from pathlib import Path
import re
from functools import reduce
from collections import Counter


def read_input() -> list[str]:
    return [line for line in Path("input").read_text().split("\n") if len(line)]


def input_to_lists(pairs: list[str]) -> tuple[list[int], list[int]]:
    left_list: list[int] = []
    right_list: list[int] = []
    whitespace = re.compile(r"\s+")
    for nums in pairs:
        left, right = [int(num) for num in whitespace.split(nums)]
        left_list.append(left)
        right_list.append(right)
    return sorted(left_list), sorted(right_list)


def part_one() -> int:
    left, right = input_to_lists(read_input())
    return reduce(lambda acc, pair: acc + abs(pair[0] - pair[1]), zip(left, right), 0)


def part_two() -> int:
    left, right = input_to_lists(read_input())
    counter = Counter(right)
    return reduce(lambda acc, num: acc + num * counter.get(num, 0), left, 0)


if __name__ == "__main__":
    print(f"{part_one()=}")
    print(f"{part_two()=}")

File: 2024/day_01/rust/src/main.rs
use std::collections::HashMap;
use std::fs;
use std::vec::Vec;

fn read_input() -> Vec<String> {
    fs::read_to_string("../input")
        .expect("Failed to read input")
        .lines()
        .filter(|line| !line.is_empty())
        .map(|line| line.trim().to_string())
        .collect()
}

fn parse_numbers(pairs: Vec<String>) -> (Vec<u32>, Vec<u32>) {
    let (mut left, mut right): (Vec<u32>, Vec<u32>) = pairs
        .into_iter()
        .map(|pair| {
            let mut nums = pair.split_whitespace();
            let left = nums
                .next()
                .expect("Expected left number")
                .parse::<u32>()
                .expect("Unable to parse left number");
            let right = nums
                .next()
                .expect("Expected right number")
                .parse::<u32>()
                .expect("Unable to parse right number");
            (left, right)
        })
        .unzip();
    left.sort_unstable();
    right.sort_unstable();
    (left, right)
}

fn part_one(left: Vec<u32>, right: Vec<u32>) -> u32 {
    left.into_iter()
        .zip(right)
        .fold(0, |acc, pair| acc + pair.0.abs_diff(pair.1))
}

fn part_two(left: Vec<u32>, right: Vec<u32>) -> u32 {
    let counter: HashMap<u32, u32> =
        right
            .into_iter()
            .fold(HashMap::with_capacity(left.len()), |mut counter, num| {
                *counter.entry(num).or_insert(0) += 1;
                counter
            });
    left.into_iter()
        .fold(0, |acc, num| acc + num * counter.get(&num).unwrap_or(&0))
}

fn main() {
    let (left, right) = parse_numbers(read_input());
    let result_1 = part_one(left.clone(), right.clone());
    println!("Part One Result: {result_1}");
    let result_2 = part_two(left, right);
    println!("Part One Result: {result_2}");
}
